restli-server/src/main/java/com/linkedin/restli/internal/server/response/BatchFinderResponseBuilder.java

```diff
@@ -195,26 +196,52 @@ private AnyRecord buildMetaData(CollectionResult<RecordTemplate, RecordTemplate>
     return null;
   }
 
-  private String getBatchParamterName(RoutingResult routingResult)
+  private String getBatchParameterName(RoutingResult routingResult)
   {
     int batchFinderCriteriaIndex = routingResult.getResourceMethod().getBatchFinderCriteriaParamIndex();
     return routingResult.getResourceMethod().getParameters().get(batchFinderCriteriaIndex).getName();
   }
 
   private DataList getCriteriaParameters(RoutingResult routingResult)
   {
-    String batchParamName = getBatchParamterName(routingResult);
-    return(DataList)routingResult.getContext().getStructuredParameter(batchParamName);
+    String batchParameterName = getBatchParameterName(routingResult);
+    return(DataList)routingResult.getContext().getStructuredParameter(batchParameterName);
   }
 
-  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String queryParam, URI uri)
+  private URI buildCriteriaURI(ResourceContextImpl resourceContext, RecordTemplate criteria, String batchParamName, URI uri)
   {
     DataList criteriaList = new DataList(1);
     criteriaList.add(criteria.data());
+    DataMap queryParams = extractQueryParamsFromResourceContext(resourceContext);
     return URIParamUtils.replaceQueryParam(uri,
-                                           queryParam,
+                                           batchParamName,
                                            criteriaList,
-                                           resourceContext.getParameters(),
+                                           queryParams,
                                            resourceContext.getRestliProtocolVersion());
   }
+
+  private DataMap extractQueryParamsFromResourceContext(ResourceContextImpl resourceContext)
+  {
+    try
+    {
+      DataMap queryParams = resourceContext.getParameters().copy();
+      if (queryParams.containsKey(RestConstants.FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.FIELDS_PARAM, resourceContext.getProjectionMask().getDataMap());
+      }
+      if (queryParams.containsKey(RestConstants.PAGING_FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.PAGING_FIELDS_PARAM, resourceContext.getPagingProjectionMask().getDataMap());
+      }
+      if (queryParams.containsKey(RestConstants.METADATA_FIELDS_PARAM))
+      {
+        queryParams.put(RestConstants.METADATA_FIELDS_PARAM, resourceContext.getMetadataProjectionMask().getDataMap());
+      }
+      return queryParams;
+    }
+    catch (CloneNotSupportedException cloneNotSupportedException)
+    {
+      return resourceContext.getParameters();
```

Comment:
This is not a good way to recover this error. I'd prefer failing the request as this is unexpected.

Reply:
Instead of `copy` I used `clone`, according to the comment and short code trace, clone should be the shallow copy